### PR TITLE
Refactor metrics

### PR DESCRIPTION
**Pull request by d4nilos**
This change is refactoring the way we handle the metrics.
A new class was created: NigtableMetric
The metric names (tags) was centralized in this class, and the code was refactored to make the flow of registering and unregistering metrics dependent only of the values setted there.
Consequently, any new metric to be added, or metric to be removed, we are forced to add in the BigtableMetric, and the flow for unregistering it is already fulfilled, you just need to add the registering logic (that was also simplified, adding the new metric in the switch case).

### CODE REVIEW

File: src/main/java/com/spotify/autoscaler/ClusterStats.java [hunk 1]
```diff
@@ -211,62 +202,70 @@ void setStats(BigtableCluster cluster, int currentNodes) {
...
-                          p ->
-                              Optional.of(
-                                  Duration.between(
-                                      p.lastCheck().orElse(Instant.EPOCH), Instant.now())))
-                      .get()
-                      .getSeconds());
+      for (BigtableMetric.Metrics metric : BigtableMetric.Metrics.values()) {
+        Gauge metricValue = null;
+        switch (metric) {
+          case NODE_COUNT:
```
Comment (gizem969) [inline src/main/java/com/spotify/autoscaler/ClusterStats.java]: maybe the enum `BigtableMetric.Metrics` can have a method that returns a lambda instead of this `switch` statement

File: src/main/java/com/spotify/autoscaler/metric/BigtableMetric.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    CPU_TARGET_RATIO("cpu-target-ratio");
+
+    public String tag;
+
+    Metrics(final String tag) {
+      this.tag = tag;
+    }
+  }
+
+  public enum MetricType {
```
Comment (gizem969) [inline src/main/java/com/spotify/autoscaler/metric/BigtableMetric.java]: this `Metrics` and `MetricType` naming is pretty weird. Maybe this one can be `LoadMetricType` or something

File: src/main/java/com/spotify/autoscaler/ClusterStats.java [hunk 3]
```diff
@@ -211,62 +107,43 @@ void setStats(BigtableCluster cluster, int currentNodes) {
...
+                                cluster.projectId(), cluster.instanceId(), cluster.clusterId())
+                            .flatMap(
+                                p ->
+                                    Optional.of(
+                                        Duration.between(
+                                            p.lastCheck().orElse(Instant.EPOCH), Instant.now())))
+                            .get()
+                            .getSeconds();
+            break;
+          case CPU_TARGET_RATIO:
```
Comment (gizem969) [inline src/main/java/com/spotify/autoscaler/ClusterStats.java]: this can be 
`return () -> clusterData.getCpuUtil() / clusterData.getCluster().cpuTarget();` in the enum

File: src/main/java/com/spotify/autoscaler/ClusterStats.java [hunk 4]
```diff
@@ -279,18 +156,6 @@ void setStats(BigtableCluster cluster, int currentNodes) {
...
-              .tagged("project-id", cluster.projectId())
-              .tagged("cluster-id", cluster.clusterId())
-              .tagged("instance-id", cluster.instanceId()),
-          (Gauge<Double>)
-              () -> {
-                final ClusterData data = registeredClusters.get(cluster.clusterName());
-                return data.getCpuUtil() / cluster.cpuTarget();
-              });
     } else {
       clusterData.setCurrentNodeCount(currentNodes);
```
Comment (gizem969) [inline src/main/java/com/spotify/autoscaler/ClusterStats.java]: this is not related to this pr, but i think there should be a `clusterData.setCluster` here, because the configs might have changed, like the `cpuTarget` for example. then the cpu_target_ratio metric above becomes wrong.

File: src/main/java/com/spotify/autoscaler/ClusterStats.java [hunk 5]
```diff
@@ -211,62 +107,43 @@ void setStats(BigtableCluster cluster, int currentNodes) {
...
-                                      p.lastCheck().orElse(Instant.EPOCH), Instant.now())))
-                      .get()
-                      .getSeconds());
+      for (BigtableMetric.Metrics metric : BigtableMetric.Metrics.values()) {
+        // This will return null for non implemented getMetricValues. This will happen when the
+        // metric depends in other things other than the registeredClusters or the cluster
+        // itself, e.g., when it is dependent on the database.
+        Gauge metricValue = metric.getMetricValue(registeredClusters.get(cluster.clusterName()));
+        switch (metric) {
+          case LAST_CHECK_TIME:
```
Comment (gizem969) [inline src/main/java/com/spotify/autoscaler/ClusterStats.java]: i don't like this here :)
i can't think of anything other than passing the `db` to the enum, but even that would be better than this i think.
any other opinions?

Comment (d4nilos) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: I don't like it neither, I'm trying to think in a better solution...
My initial approach leave the metrics implementation in the ClusterStats was to avoid sending too many contextual objects to the Metrics class... :( 

Still thinking on how to improve this... :)

Comment (d4nilos) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: I added the db, I'm not feeling so bad as I thought I would feel with this change... :)

File: src/main/java/com/spotify/autoscaler/ClusterStats.java (hunk 4 above)
Comment (Bj0rnen) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: I'm finding this aspect hard to follow. I might be misreading the current code, so this might not make sense:

My opinion is that we should confine all shared mutable state down to the one `ConcurrentHashMap<String, ClusterData>` of _immutable_ `ClusterData` objects. The lambdas shouldn't capture anything but the key to this map, so for example not the `cluster.cpuTarget()`. `setStats` shouldn't lookup a `ClusterData` and mutate it; it should replace it with a new `ClusterData`. Once one of these lambdas looks up the `ClusterData` in the map, that should contain all the information that is needed to report a metric.

Comment (gizem969) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: yeah that sounds good too. then the `putIfAbsent` should be replaced with a `put`, and also the `setLoad` method should change to sth similar.

File: src/main/java/com/spotify/autoscaler/ClusterStats.java (hunk 3 above)
Comment (d4nilos) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: 👍

File: src/main/java/com/spotify/autoscaler/ClusterStats.java (hunk 4 above)
Comment (d4nilos) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: I see your (scio) point here, but do you feel that we should add this change in this PR?

File: src/main/java/com/spotify/autoscaler/ClusterStats.java [hunk 6]
```diff
@@ -211,62 +105,20 @@ void setStats(BigtableCluster cluster, int currentNodes) {
...
-                          cluster.projectId(), cluster.instanceId(), cluster.clusterId())
-                      .flatMap(
-                          p ->
-                              Optional.of(
-                                  Duration.between(
-                                      p.lastCheck().orElse(Instant.EPOCH), Instant.now())))
-                      .get()
-                      .getSeconds());
+      for (BigtableMetric.Metrics metric : BigtableMetric.Metrics.values()) {
+        // This will return null for non implemented getMetricValues. This will happen when the
```
Comment (gizem969) [inline src/main/java/com/spotify/autoscaler/ClusterStats.java]: this is outdated now

File: src/main/java/com/spotify/autoscaler/ClusterStats.java (hunk 4 above)
Comment (gizem969) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: I think we can simply add `clusterData.setCluster` here, and do the bigger refactor in another pr. wdyt @Bj0rnen ?

Comment (d4nilos) [reply src/main/java/com/spotify/autoscaler/ClusterStats.java]: I added the setCluster, let's discuss this better tomorrow and not merge it today :)

File: src/main/java/com/spotify/autoscaler/metric/BigtableMetric.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+      }
+    };
+
+    public String tag;
+
+    Metrics(final String tag) {
+      this.tag = tag;
+    }
+
+    public Gauge getMetricValue(final ClusterData clusterData, final Database db) {
```
Comment (gizem969) [inline src/main/java/com/spotify/autoscaler/metric/BigtableMetric.java]: sorry this is my fault, but can you change it to `abstract` 👼 
`public abstract Gauge getMetricValue(final ClusterData clusterData) ;`
